src/lib.rs

```diff
@@ -833,6 +855,25 @@ where
         self
     }
 
+    ///
+    /// Sets the introspection URL for contacting the ([RFC 7662](https://tools.ietf.org/html/rfc7662))
+    /// introspection endpoint.
+    ///
+    pub fn set_introspection_url(mut self, introspection_url: IntrospectionUrl) -> Self {
+        self.oauth2_client = self.oauth2_client.set_introspection_url(introspection_url);
+        self
+    }
+
+    ///
+    /// Sets the revocation URL for contacting the revocation endpoint ([RFC 7009](https://tools.ietf.org/html/rfc7009)).
+    ///
+    /// See: [`revoke_token()`](Self::revoke_token())
+    ///
+    pub fn set_revocation_uri(mut self, revocation_uri: RevocationUrl) -> Self {
```

Comment:
thoughts on `uri` vs `url` here?

Reply:
I copied the existing pattern set by `fn set_redirect_uri()` although now I see that `fn set_revocation_url()` is inconsistent. RFC 6749 refers to redirection URI while RFC 7009 refers to revocation URL. RFC 7662 ducks the issue and refers only to the token endpoint defined by RFC 6749 which gives us 2/3 for URI and 1/3 for URL. URI can be either a URL or a URN, but in our case we are only talking about URLs, not URNs, because clients have to make HTTP requests to these endpoints and they can only make those requests to the URL subset of URIs, or even a subset of URLs (e.g. HTTPS and not HTTP). According to https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata the authorization_endpoint, token_endpoint and userinfo_endpoint are all URLs, not URIs.

I'd be inclined to standardize on URL.

Reply:
Regarding this:

> ah, I had a thought about the delayed error when reviewing the oauth2 change but forgot about the NoUserInfoEndpoint trick I used in this crate. I think NoUserInfoEndpoint is probably a better pattern since the user gets the error sooner, but either approach seems ok.

I think the NoUserInfoEndpoint approach is preferable as it not only gives earlier feedback to the developer but also the TokenRequestError based approach "hides" an input error in among many other possible types of error making it harder to take appropriate action in the calling code.

Should I therefore push a change to the oauth2-rs crate to adopt this style of error handling for the revocation and introspection URLs, adding errors for missing and invalid endpoint, and going further do we even need separate errors per endpoint (e.g. NoUserInfoEndpoint, NoIntrospectionEndpoint, NoRevocationEndpoint) or is it enough to just re-use an error such as MissingEndpointURL and InvalidEndpointURL ?

Reply:
> Should I therefore push a change to the oauth2-rs crate to adopt this style of error handling for the revocation and introspection URLs, adding errors for missing and invalid endpoint

if you'd like to, please go ahead! I'll happily merge this improvement.

> and going further do we even need separate errors per endpoint (e.g. NoUserInfoEndpoint, NoIntrospectionEndpoint, NoRevocationEndpoint) or is it enough to just re-use an error such as MissingEndpointURL and InvalidEndpointURL ?

using a common/shared error type seems fine. I'd suggest including a `&'static str` within it so that the `Display` output says which specific endpoint URL is missing, but I don't see any reason we need to differentiate between those errors in terms of static typing

Reply:
> I'd be inclined to standardize on URL.

sgtm